src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node head;
+    private Node tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        addLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            Node prev = getNodeByIndex(index);
+            insertBefore(prev, value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node nodeToSet = getNodeByIndex(index);
+        T oldValue = nodeToSet.element;
+        nodeToSet.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToRemove = getNodeByIndex(index);
+        return unlinkNode(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        while (currentNode != null) {
+            if (object == currentNode.element || object != null
+                    && object.equals(currentNode.element)) {
+                unlinkNode(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node {
+        private T element;
+        private Node prev;
+        private Node next;
+
+        private Node(T element, Node prev, Node next) {
+            this.element = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bounds, size " + size);
+        }
+    }
+
+    private void addFirst(T elem) {
+        Node newNode = new Node(elem, null, head);
+        Node oldNode = head;
+        head = newNode;
+        if (oldNode == null) {
+            tail = newNode;
+        } else {
+            oldNode.prev = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T elem) {
+        Node newNode = new Node(elem, tail, null);
+        Node oldNode = tail;
+        tail = newNode;
+        if (oldNode == null) {
+            head = newNode;
+        } else {
+            oldNode.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlinkNode(Node nodeToUnlink) {
```

Comment:
Look throu your code and parametrize node everywhere
```suggestion
    private T unlinkNode(Node<T> nodeToUnlink) {
```